lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,341 @@
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define NRF_INBUILT_KEY_OP_LS "AT%CMNG=1"
+#define NRF_INBUILT_KEY_OP_RD "AT%CMNG=2"
+#define NRF_INBUILT_KEY_OP_WR "AT%CMNG=0"
+#define NRF_INBUILT_KEY_OP_RM "AT%CMNG=3"
+
+#define NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 (10)
+#define NRF_INBUILT_KEY_CR_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_LF_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_QUOTES_CHAR_SIZE (1)
+#define NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE (1)
+#define NRF_INBUILT_KEY_COMMA_SIZE (1)
+#define NRF_INBUILT_KEY_INVALID_HANDLE (-1)
+
+static u8_t scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	enum at_cmd_state state;
+
+	int retval =
+		at_cmd_write("AT+CMEE?", response, sizeof(response), &state);
+	if (retval == 0) {
+		char *deactivated = strchr(&response[6], '0');
+
+		if (deactivated != NULL) {
+			return 0;
+		}
+
+		char *activated = strchr(&response[6], '1');
+
+		if (activated != NULL) {
+			return 1;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	char response[10];
+	enum at_cmd_state state;
+
+	int retval = at_cmd_write(at_cmee_strings[(int)enable], response,
+				  sizeof(response), &state);
+
+	if (retval < 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+int nrf_inbuilt_key_write(nrf_sec_tag_t sec_tag,
+			  nrf_key_mgnt_cred_type_t cred_type, u8_t *p_buffer,
+			  u16_t buffer_len)
+{
+	if ((p_buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_WR) + buffer_len +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 3) +
+				(NRF_INBUILT_KEY_QUOTES_CHAR_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	u8_t *p_cmd = &scratch_buf[0];
+
+	memset(p_cmd, 0, scratch_buf_len);
+
+	int exp_len = snprintf((char *)p_cmd, scratch_buf_len, "%s,%u,%u,\"",
+			       NRF_INBUILT_KEY_OP_WR, (u32_t)sec_tag,
+			       (u8_t)cred_type);
+
+	memcpy(&p_cmd[exp_len], p_buffer, buffer_len);
+	exp_len += buffer_len;
+
+	memcpy(&p_cmd[exp_len], "\"\r\n", 4);
+	exp_len += 4;
+
+	if (exp_len >= scratch_buf_len) {
+		return -ENOBUFS;
+	}
+
+	size_t expected_size = strlen((char *)p_cmd);
+
+	expected_size +=
+		NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	bool cmee_activated = false;
+	int cmee_active_result = cmee_active();
+
+	if (cmee_active_result < 0) {
+		return -EFAULT;
+	}
+
+	if (cmee_active_result == 0) {
+		cmee_set(true);
+		cmee_activated = true;
+	}
+
+	enum at_cmd_state state;
+	int result =
+		at_cmd_write(scratch_buf, scratch_buf, scratch_buf_len, &state);
```

Comment:
It seems risky to use the same buffer for input and output. And anyways, we don't read the response do we? So perhaps we should use NULL ?

Reply:
That's actually true. The CME error result is now returned in the positive side at_cmd retval. Good point. We do not need the read buffer in the write/delete function.